File: backend/src/main/java/com/politeai/infrastructure/ai/preprocessing/LockedSpanMasker.java
package com.politeai.infrastructure.ai.preprocessing;

import com.politeai.domain.transform.model.LockedSpan;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Masks locked spans with placeholders before sending to LLM,
 * and unmasks (restores) them in the LLM output.
 */
@Slf4j
@Component
public class LockedSpanMasker {

    // Flexible pattern for matching type-specific placeholders in LLM output
    // Handles variations: {{DATE_1}}, {{ DATE_1 }}, {{DATE-1}}, etc.
    // Captures group 1: prefix (e.g., DATE), group 2: counter (e.g., 1)
    private static final Pattern PLACEHOLDER_PATTERN = Pattern.compile(
            "\\{\\{\\s*([A-Z]+)[-_](\\d+)\\s*\\}\\}"
    );

    /**
     * Replace locked spans in the original text with their placeholders.
     * Spans must be sorted by startPos ascending.
     *
     * @param text  the original text
     * @param spans the extracted locked spans (sorted by position)
     * @return text with locked spans replaced by placeholders
     */
    public String mask(String text, List<LockedSpan> spans) {
        if (spans == null || spans.isEmpty()) {
            return text;
        }

        StringBuilder sb = new StringBuilder();
        int lastEnd = 0;

        for (LockedSpan span : spans) {
            // Append text before this span
            sb.append(text, lastEnd, span.startPos());
            // Append placeholder
            sb.append(span.placeholder());
            lastEnd = span.endPos();
        }

        // Append remaining text
        sb.append(text, lastEnd, text.length());

        return sb.toString();
    }

    /**
     * Restore placeholders in the LLM output with their original text.
     * Uses flexible matching to handle minor LLM variations in placeholder format.
     *
     * @param output the LLM output text containing placeholders
     * @param spans  the original locked spans
     * @return text with placeholders restored to original values, and count of missing spans
     */
    public UnmaskResult unmask(String output, List<LockedSpan> spans) {
        if (spans == null || spans.isEmpty()) {
            return new UnmaskResult(output, List.of());
        }

        // Build placeholder→span map using canonical placeholder string (e.g., "{{DATE_1}}")
        java.util.Map<String, LockedSpan> spanMap = new java.util.HashMap<>();
        java.util.Set<String> restored = new java.util.HashSet<>();
        for (LockedSpan span : spans) {
            spanMap.put(span.placeholder(), span);
        }

        // Replace all placeholders found in the output
        String result = output;
        Matcher matcher = PLACEHOLDER_PATTERN.matcher(result);
        StringBuilder sb = new StringBuilder();
        while (matcher.find()) {
            String prefix = matcher.group(1);
            String counter = matcher.group(2);
            String canonical = "{{" + prefix + "_" + counter + "}}";
            LockedSpan span = spanMap.get(canonical);
            if (span != null) {
                matcher.appendReplacement(sb, Matcher.quoteReplacement(span.originalText()));
                restored.add(canonical);
            } else {
                log.warn("LockedSpan placeholder {} not found in span map", canonical);
            }
        }
        matcher.appendTail(sb);
        result = sb.toString();

        // Check for missing spans
        List<LockedSpan> missingSpans = new java.util.ArrayList<>();
        for (LockedSpan span : spans) {
            if (!restored.contains(span.placeholder())) {
                log.warn("LockedSpan missing in output: placeholder={}, type={}, text='{}'",
                        span.placeholder(), span.type(), span.originalText());

                // Check if the original text appears verbatim in the output (LLM may have expanded it)
                if (!result.contains(span.originalText())) {
                    missingSpans.add(span);
                } else {
                    log.info("LockedSpan {} found as verbatim text in output (LLM preserved without placeholder)", span.placeholder());
                }
            }
        }

        return new UnmaskResult(result, missingSpans);
    }

    /**
     * Result of the unmask operation.
     *
     * @param text         the unmasked text
     * @param missingSpans spans whose placeholders were not found and original text is also absent
     */
    public record UnmaskResult(String text, List<LockedSpan> missingSpans) {}
}

File: app/pipeline/preprocessing/locked_span_masker.py
import logging
import re
from dataclasses import dataclass

from app.models.domain import LockedSpan

logger = logging.getLogger(__name__)

# Flexible pattern for matching type-specific placeholders in LLM output
# Handles variations: {{DATE_1}}, {{ DATE_1 }}, {{DATE-1}}, etc.
_PLACEHOLDER_PATTERN = re.compile(r"\{\{\s*([A-Z]+)[-_](\d+)\s*\}\}")


@dataclass
class UnmaskResult:
    text: str
    missing_spans: list[LockedSpan]


def mask(text: str, spans: list[LockedSpan]) -> str:
    """Replace locked spans in the original text with their placeholders.

    Spans must be sorted by start_pos ascending.
    """
    if not spans:
        return text

    parts: list[str] = []
    last_end = 0

    for span in spans:
        parts.append(text[last_end:span.start_pos])
        parts.append(span.placeholder)
        last_end = span.end_pos

    parts.append(text[last_end:])
    return "".join(parts)


def unmask(output: str, spans: list[LockedSpan]) -> UnmaskResult:
    """Restore placeholders in the LLM output with their original text.

    Uses flexible matching to handle minor LLM variations in placeholder format.
    """
    if not spans:
        return UnmaskResult(text=output, missing_spans=[])

    # Build placeholder -> span map
    span_map: dict[str, LockedSpan] = {}
    for span in spans:
        span_map[span.placeholder] = span

    restored: set[str] = set()

    def replace_fn(m: re.Match) -> str:
        prefix = m.group(1)
        counter = m.group(2)
        canonical = f"{{{{{prefix}_{counter}}}}}"
        span = span_map.get(canonical)
        if span is not None:
            restored.add(canonical)
            return span.original_text
        logger.warning("LockedSpan placeholder %s not found in span map", canonical)
        return m.group(0)

    result = _PLACEHOLDER_PATTERN.sub(replace_fn, output)

    # Check for missing spans
    missing_spans: list[LockedSpan] = []
    for span in spans:
        if span.placeholder not in restored:
            logger.warning(
                "LockedSpan missing in output: placeholder=%s, type=%s, text='%s'",
                span.placeholder,
                span.type,
                span.original_text,
            )
            if span.original_text not in result:
                missing_spans.append(span)
            else:
                logger.info(
                    "LockedSpan %s found as verbatim text in output (LLM preserved without placeholder)",
                    span.placeholder,
                )

    return UnmaskResult(text=result, missing_spans=missing_spans)
